compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp

```diff
@@ -239,72 +296,149 @@ void ReductionStrategy3StageThreadDistribution::compute(
     else
       vectorSizeStage2 = 1;
   }
+
+  // Stage 3
+  // -------
+  // Always happens by default with warp shuffles.
 }
 
-/// Structure to hold the parameters related to GPU reduction strategy.
-struct GPUReductionStrategyInfos {
-  explicit GPUReductionStrategyInfos(
-      MLIRContext *context, transform_ext::MatchedReductionCaptures captures)
-      : context(context),
-        reductionRank(captures.reductionRank),
-        reductionDimensionSize(captures.reductionDimensionSize),
-        maybeLeadingRank(captures.maybeLeadingRank),
-        maybeTrailingRank(captures.maybeTrailingRank) {
-    auto blockX =
-        mlir::gpu::GPUBlockMappingAttr::get(context, mlir::gpu::Blocks::DimX);
-    auto blockY =
-        mlir::gpu::GPUBlockMappingAttr::get(context, mlir::gpu::Blocks::DimY);
-    auto blockZ =
-        mlir::gpu::GPUBlockMappingAttr::get(context, mlir::gpu::Blocks::DimZ);
-    allBlockAttrs = SmallVector<Attribute>{blockX, blockY, blockZ};
+/// Encodes a strategy targeted at (very) small reductions.
+///
+/// In the case of small reductions, we cannot make an efficient use of warp
+/// shuffles. Instead, try to take advantage of caches.
+/// This strategy aims at running the reduction sequentially within each thread
+/// and taking parallelism from outer dimensions that we would otherwise use
+/// for block-level parallelism.
+///
+/// There are 2 cases:
+///   1. we can find good divisors of outer parallel dimensions and avoid
+///      creating dynamic tile sizes. We can then vectorize to the reduction
+///      size. Empirically, this works well up to reduction sizes of ~2-3 warps,
+///      so we filter static dimensions up to 2*warp_size - 1 for now.
+///   2. we cannot find good divisors, we pay the price of dynamic loops and
+///      avoid catastrophic performance cliffs. Empirically, this works well
+///      up to reduction sizes of ~1 warp.
+///
```

Comment:
Add a note: the ability of this strategy to run at peak is highly dependent on properly using vector<4/2>, this currently happens only on size multiples of 4/2. This has been corroborated by looking at perf dips on uneven sizes and inspecting the ptx.